File: haskell/test/SerializeSpec.hs
module SerializeSpec (spec) where

import Data.Aeson (encode)
import qualified Data.ByteString.Lazy.Char8 as LBS
import Test.Hspec
import TestTypes

-- Interop JSON — identical strings used across all 12 language libraries
circleJson, rectangleJson, carJson, truckJson :: String
circleJson = "{\"#type\":\"Circle\",\"radius\":4}"
rectangleJson = "{\"#type\":\"Rectangle\",\"width\":10,\"height\":5}"
carJson = "{\"#type\":\"Car\",\"make\":\"Toyota\"}"
truckJson = "{\"#type\":\"Truck\",\"payloadTons\":5}"

-- Aeson sorts fields alphabetically, so Rectangle comes out as height,width
aesonRectangleJson :: String
aesonRectangleJson = "{\"#type\":\"Rectangle\",\"height\":5,\"width\":10}"

spec :: Spec
spec = describe "serialize" $ do

  it "encode_whenSerializingCircle_matchesInteropFormat" $ do
    -- Act
    let json = LBS.unpack (encode (Circle { radius = 4.0 }))
    -- Assert
    json `shouldBe` circleJson

  it "encode_whenSerializingRectangle_matchesInteropFormat" $ do
    -- Act
    let json = LBS.unpack (encode (Rectangle { width = 10.0, height = 5.0 }))
    -- Assert
    json `shouldBe` aesonRectangleJson

  it "encode_whenSerializingCar_matchesInteropFormat" $ do
    -- Act
    let json = LBS.unpack (encode (PersonalVehicle { make = "Toyota" }))
    -- Assert
    json `shouldBe` carJson

  it "encode_whenSerializingTruck_matchesInteropFormat" $ do
    -- Act
    let json = LBS.unpack (encode (TransportationVehicle { payloadTons = 5.0 }))
    -- Assert
    json `shouldBe` truckJson

File: scala/src/test/scala/jsontag/SerializeSpec.scala
package jsontag

import munit.FunSuite

class SerializeSpec extends FunSuite:

  // Interop JSON — identical strings used across all 12 language libraries
  val circleJson = """{"#type":"Circle","radius":4}"""
  val rectangleJson = """{"#type":"Rectangle","width":10,"height":5}"""
  val carJson = """{"#type":"Car","make":"Toyota"}"""
  val truckJson = """{"#type":"Truck","payloadTons":5}"""

  test("serialize, when serializing Circle, matches interop format"):
    // Act
    val json = JsonTag.write(Circle(4.0): Shape)

    // Assert
    assertEquals(json, circleJson)

  test("serialize, when serializing Rectangle, matches interop format"):
    // Act
    val json = JsonTag.write(Rectangle(10.0, 5.0): Shape)

    // Assert
    assertEquals(json, rectangleJson)

  test("serialize, when serializing Car, matches interop format"):
    // Act
    val json = JsonTag.write(PersonalVehicle("Toyota"): Vehicle)

    // Assert
    assertEquals(json, carJson)

  test("serialize, when serializing Truck, matches interop format"):
    // Act
    val json = JsonTag.write(TransportationVehicle(5.0): Vehicle)

    // Assert
    assertEquals(json, truckJson)
